pkg/ipam/poolallocator/testing/fake_client.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testing
+
+import (
+	"sync"
+
+	"github.com/google/uuid"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	k8stesting "k8s.io/client-go/testing"
+
+	crdv1a2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	fakeversioned "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+)
+
+// Simple client is not sufficient for pool allocator testing,
+// since pool allocator relies on both crd client and pool informer
+// to work in sync. This client extension mimics the real client in
+// conflict handling functionality - pool update will return conflict
+// error unless APIVersion for the updated pool reflect the version
+// stored in the client.
```

Comment:
crdClient and pool informer can work in sync: objects updated by crdClient will be reflected in informer, what's the problem here?
If you want to mimic conflict, you could add a reactor to crdClient directly, for example:
```
updateCalled := 0
fakeClient.AddReactor("update", "*", func(action k8stesting.Action) (bool, runtime.Object, error) {
	updateCalled += 1
	if updateCalled <= 1 {
		return true, nil, ConflictError
	}
	...
})
```

Reply:
The problem is in synchronization, informer is async and its not updated right away. 
For example, when first IP is allocated, it its saved in pool state, however this update is not yet propagated to the informer when second IP allocation is ready to be saved. 
Unless conflict is detected as shown in the code below, such scenario results in second allocation overriding the first one:
https://github.com/antrea-io/antrea/blob/main/pkg/ipam/poolallocator/allocator.go#L180

Instead of conflict being an edge test case, rather every test that allocates multiple IPs involves a conflict, so it appears we need extra tooling to enable these tests.

Reply:
resourceVersion is designed to solve the problem exactly. For a real apiserver, you will get conflict error if the resourceVersion in the request doesn't match the one stored in apiserver, and each update will increase the resourceVersion.

So maybe you still don't need to create this client struct with the extra storage. Instead, you could check whether the resourceVersion of the updated object matches the one got via `crdClient.CrdV1alpha2().ExternalIPPools().Get`. You can prepend a reactor to ippools's update method which generates a new resourceVersion and return `handled` as false so it will continue to store the updated object.

Reply:
I think a dedicated testing client is a good idea since it is used in multiple places (pool allocator, antrea ipam driver). If you agree, then I think the only improvement you are suggesting is getting rid of pool storage in the testing client, while replacing it with a `Get` call from the same client?

Reply:
Yes, I'm only suggesting to get rid of the extra storage and extra code that make it work but not easy to understand why they are needed, i.e. L52-L54, L73, to reduce the maintainance efforts, when the fakeClient itself is flexible enough to handle this situation.
I think you can still use a function to create an IPPoolClient for testing, in which some reactors will be added.

Reply:
I don't think this can work easily, since there's a lock in fake_client that prevents us from calling `get` from within `update`:
https://github.com/kubernetes/client-go/blob/ad95a77727e65f037cc877600f4d8ea4d867c617/testing/fixture.go#L201

I have reduced the storage in test client to contain `ResourceVersion` only, rather than the whole pool. I also feel that storing `watcher` on the client is easier and more readable than retrieving it by resource name from watcher list. Please let me know if you agree or rather there is a solution that I overlooked! Thanks

Reply:
Do we have any more progress about this? Otherwise we can approve and merge this PR.

Reply:
Unless @tnqn has any comments, I think we can merge